modules/pipeline/services/crondsvc/crond.go

```diff
@@ -14,43 +14,98 @@
 package crondsvc
 
 import (
-	"context"
 	"fmt"
+	"strconv"
+	"strings"
 
 	"github.com/sirupsen/logrus"
 
 	"github.com/erda-project/erda/modules/pipeline/conf"
-	"github.com/erda-project/erda/modules/pipeline/spec"
 	"github.com/erda-project/erda/pkg/cron"
-	"github.com/erda-project/erda/pkg/jsonstore/storetypes"
+	"github.com/erda-project/erda/pkg/strutil"
 )
 
-const etcdCrondUpdateWatchKey = "/devops/pipeline/crond/update"
+const cronChanPrefixDeleteKey = "delete-"
+const cronChanPrefixAddKey = "add-"
 
+// Listen the key in the channel
+// decide to delete or add a scheduled task according to the prefix of the key
 func (s *CrondSvc) ListenCrond(pipelineCronFunc func(id uint64)) {
-	logrus.Info("crond: start listen")
 	for {
-		// watch: isPrefix[true], filterDelete[true], keyOnly[true]
-		err := s.js.IncludeWatch().Watch(context.Background(), etcdCrondUpdateWatchKey, false, true, true, nil,
-			func(key string, _ interface{}, t storetypes.ChangeType) error {
-				// sync update
-				logrus.Infof("crond: watched update request key, key: %s", key)
-				if _, err := s.ReloadCrond(pipelineCronFunc); err != nil {
-					logrus.Errorf("[alert] crond: failed to reload crond (%v)", err)
-					return nil
+		select {
+		case key := <-s.cronChan:
+			logrus.Infof("crond: watched update chan key: %s", key)
+
+			cronID, err := parseCronIDFromWatchedKey(key)
+			if err != nil {
+				logrus.Errorf("crond: failed to parse cronID from watched key, key: %s, err: %v", key, err)
+				continue
+			}
+
+			if strings.HasPrefix(key, cronChanPrefixAddKey) {
+				pc, err := s.dbClient.GetPipelineCron(cronID)
+				if err != nil {
+					logrus.Errorf("crond: failed to get cron cronID: %v error: %v", cronID, err)
+					continue
+				}
+				// why delete it first, because crond.AddFunc cannot add a scheduled task with the same name
+				err = s.crond.Remove(makePipelineCronName(cronID))
+				if err != nil {
+					logrus.Errorf("crond: failed to remove cron cronID: %v error: %v", cronID, err)
+					continue
 				}
-				logrus.Infof("crond: watched and reload successfully")
-				return nil
-			})
-		if err != nil {
-			logrus.Errorf("[alert] crond: watch failed, err: %v", err)
+
+				// determine whether there is a scheduled task
+				if pc.Enable != nil && *pc.Enable && pc.CronExpr != "" {
```

Comment:
judge before send to channel is better?

Reply:
add cron was get pipeline cron from db, So it is necessary to determine whether the latest value is open